src/Hazelcast.Net.Linq.Async/Visitors/QueryBinder.cs

```diff
@@ -0,0 +1,283 @@
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using System.Text;
+using Hazelcast.Linq.Expressions;
+
+namespace Hazelcast.Linq.Visitors
+{
+    /// <summary>
+    /// Traverses and prepares a SQLized expression tree to be traversed and converted to text based SQL statements.
+    /// </summary>
+    internal class QueryBinder : ExpressionVisitor
+    {
+        private ColumnProjector _projector;
+        private Dictionary<ParameterExpression, Expression> _map;
+        private int _aliasCount;
+        private const BindingFlags bindingFlags = BindingFlags.Public | BindingFlags.Instance;
+
+        public QueryBinder()
+        {
+            _projector = new ColumnProjector(p => p.NodeType == (ExpressionType)HzExpressionType.Column);
+        }
+
+        /// <summary>
+        /// Get an SQLized tree and its projection bindings
+        /// </summary>
+        /// <param name="expression"></param>
+        /// <returns></returns>
+        public Expression Bind(Expression expression)
+        {
+            _map = new Dictionary<ParameterExpression, Expression>();
+            return Visit(expression);
+        }
+
+        ///(Internal for tests)
+        internal static Expression StripQuotes(Expression expression)
+        {
+            while (expression.NodeType == ExpressionType.Quote)
+                expression = ((UnaryExpression)expression).Operand;
+
+            return expression;
+        }
+
+        /// <summary>
+        /// Generates an alias for the field. (Internal for tests)
+        /// </summary>
+        /// <returns>alias</returns>
+        internal string GetNextAlias()
+        {
+            return "t" + (_aliasCount++);
+        }
+
+        private static LambdaExpression GetLambda(Expression e)
+        {
+            e = StripQuotes(e);
+
+            if (e.NodeType == ExpressionType.Constant)
+                return ((ConstantExpression)e).Value as LambdaExpression;
+
+            return e as LambdaExpression;
+        }
+
+        public ProjectedColumns Project(Expression expression, string newAlias, string existingAlias)
+        {
+            return _projector.Project(expression, newAlias, existingAlias);
+        }
+
+        protected override Expression VisitMethodCall(MethodCallExpression m)
+        {
+            if (m.Method.DeclaringType == typeof(Queryable) || m.Method.DeclaringType == typeof(Enumerable))
+            {
+                switch (m.Method.Name)
+                {
+                    case "Where":
+                        //the arguments respectivly->type of entry, source e, predicate
+                        return this.BindWhere(m.Type, m.Arguments[0], GetLambda(m.Arguments[1]));
+                }
+
+                throw new NotSupportedException(string.Format("The method '{0}' is not supported", m.Method.Name));
```

Comment:
Being pedant here, but: all exception messages traditionally end with a dot in .NET.
Can you check them all?

Reply:
And also, we try to use interpolated strings everywhere, so
`NotSupportedException($"The method '{m.Method.Name}' is not supported"))`